link_shortener/commands/update.py

```diff
@@ -35,6 +36,8 @@ async def update_link(request, link_id, data):
             await trans.close()
             raise NotFoundException
 
+        await endpoint_duplicity_check(conn, trans, data)
```

Comment:
Tak tady by to taky šlo. Prostě check co nejdřív a transakci hned za něj. Snad to nic nerozbije. Zjednoduší se ta nová funkce

Reply:
fixnute
7aef7bebd4361c4c9465efdf8bcb823f54f59285

Reply:
V pripade` update` ale musime tento check posunt dalej a pridat check na porovani sucasneho endpointu s tym ktory sme poslali. Teraz ked user bude updetovat link bez zmeny endpointu, tak mu vychodi duplicate error.
No ale aby porovnat endpointy, musime si tie data ziskat z DB, pak ked budeme mat error, tu trans musime zavreit aby nam rasilo spravny error.
```
async def update_link(request, link_id, data):
    async with request.app.engine.acquire() as conn:
        trans = await conn.begin()
        link_update = links.update().where(links.columns['id'] == link_id)
        query = await conn.execute(links.select().where(
            links.columns['id'] == link_id
        ))
        link_data = await query.fetchone()
        if not link_data:
            await trans.close()
            raise NotFoundException

        if data.get('endpoint', None) and link_data['endpoint'] != data['endpoint']:
            await endpoint_duplicity_check(conn, data, trans)
```

```
async def endpoint_duplicity_check(conn, data, trans=None):
    query = await conn.execute(links.select().where(and_(
        links.columns['endpoint'] == data['endpoint'],
        links.columns['is_active'].is_(True)
    )))
    link_data = await query.fetchone()
    if link_data:
        if trans:
            await trans.close()
        raise DuplicateActiveLinkForbidden
```

Reply:
Transakce je potřeba jen když něco updatuješ nebo insertuješ. Na select není potřeba, takže to klidně udělej stejně jako u create a conn.begin() posuň až za check